sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/directconnectivity/ConnectionStateListenerTest.java

```diff
@@ -0,0 +1,96 @@
+import com.azure.cosmos.implementation.directconnectivity.TcpServerMock.RequestResponseType;
+import com.azure.cosmos.implementation.directconnectivity.TcpServerMock.SslContextUtils;
+import com.azure.cosmos.implementation.routing.PartitionKeyRangeIdentity;
+import io.netty.handler.ssl.SslContext;
+import org.mockito.Mockito;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.HashMap;
+import java.util.UUID;
+
+import static com.azure.cosmos.implementation.TestUtils.mockDiagnosticsClientContext;
+
+public class ConnectionStateListenerTest {
+
+    private static int port = 8082;
+    private static String serverUriString = "rntbd://localhost:" + port;
+
+    @DataProvider(name = "connectionStateListenerConfigProvider")
+    public Object[][] connectionStateListenerConfigProvider() {
+        return new Object[][]{
+            // isTcpConnectionEndpointRediscoveryEnabled, serverResponseType, GlobalAddressResolver.updateAddresses() called times
+            {true, RequestResponseType.CHANNEL_FIN, 1},
+            {false, RequestResponseType.CHANNEL_FIN, 0},
+            {true, RequestResponseType.CHANNEL_RST, 0},
+            {false, RequestResponseType.CHANNEL_RST, 0},
+        };
+    }
+
+    @Test(dataProvider = "connectionStateListenerConfigProvider")
+    public void connectionStateListener_OnConnectionEvent(boolean isTcpConnectionEndpointRediscoveryEnabled, RequestResponseType responseType, int times) {
+        TcpServer server = TcpServerFactory.startNewRntbdServer(port);
+        // Inject fake response
+        server.injectServerResponse(responseType);
+
+        ConnectionPolicy connectionPolicy = new ConnectionPolicy(DirectConnectionConfig.getDefaultConfig());
+        connectionPolicy.setTcpConnectionEndpointRediscoveryEnabled(isTcpConnectionEndpointRediscoveryEnabled);
+
+        GlobalAddressResolver addressResolver = Mockito.mock(GlobalAddressResolver.class);
+
+        SslContext sslContext = SslContextUtils.CreateSslContext("client.jks", false);
+
+        Configs config = Mockito.mock(Configs.class);
+        Mockito.doReturn(sslContext).when(config).getSslContext();
+
+        RntbdTransportClient client = new RntbdTransportClient(
+            config,
+            connectionPolicy,
+            new UserAgentContainer(),
+            addressResolver);
+
+        RxDocumentServiceRequest req =
+            RxDocumentServiceRequest.create(mockDiagnosticsClientContext(), OperationType.Create, ResourceType.Document,
+                "dbs/fakedb/colls/fakeColls",
+                getDocumentDefinition(), new HashMap<>());
+        req.setPartitionKeyRangeIdentity(new PartitionKeyRangeIdentity("fakeCollectionId","fakePartitionKeyRangeId"));
+
+        Uri targetUri = new Uri(serverUriString);
+        try {
+            client.invokeStoreAsync(targetUri, req).block();
+        } catch (Exception e) {
+            e.printStackTrace();
```

Comment:
please try to use logger here and everywhere else instead of e.printStackTrace()

Reply:
Changed to use logger here and other places.